File: Samples/Android-Advanced/app/src/main/java/com/pushwoosh/sample/inbox/InboxNotificationFactory.java
package com.pushwoosh.sample.inbox;

import android.app.Notification;
import android.app.PendingIntent;
import android.support.v4.app.NotificationCompat;
import android.support.v4.app.NotificationManagerCompat;

import com.pushwoosh.notification.PushMessage;
import com.pushwoosh.notification.PushwooshNotificationFactory;

import java.util.ArrayList;

public class InboxNotificationFactory extends PushwooshNotificationFactory {
    private static final String GROUP_KEY = "PushwooshGroup";

    private NotificationCompat.Style generateInboxStyle(PushMessage pushData) {
        MessageStorage.addMessage(getApplicationContext(), new Message(pushData.getMessage(), System.currentTimeMillis(), pushData.getHeader()));
        ArrayList<Message> messages = MessageStorage.getHistory(getApplicationContext());

        NotificationCompat.InboxStyle style = new NotificationCompat.InboxStyle().
                setBigContentTitle(pushData.getHeader() + " Details");

        for (Message message : messages) {
            style.addLine(getContentFromHtml("<b>" + message.getSender() + "</b> " + message.getText()));
        }

        if (messages.size() > 7) {
            style.setSummaryText("+ " + (messages.size() - 7) + " more");
        }

        return style;
    }

    private void generateSummary(PushMessage pushData) {
        NotificationCompat.Builder notificationBuilder = new NotificationCompat.Builder(getApplicationContext())
                .setContentTitle(getContentFromHtml(pushData.getHeader()))
                .setContentText(getContentFromHtml(pushData.getMessage()))
                .setSmallIcon(pushData.getSmallIcon())
                .setTicker(getContentFromHtml(pushData.getTicker()))
                .setWhen(System.currentTimeMillis())
                .setStyle(generateInboxStyle(pushData))
                .setGroup(GROUP_KEY)
                .setGroupSummary(true)
                .setAutoCancel(true)
                .setContentIntent(PendingIntent.getActivity(getApplicationContext(), 0, getNotificationIntent(pushData), PendingIntent.FLAG_CANCEL_CURRENT));

        NotificationManagerCompat notificationManager = NotificationManagerCompat.from(getApplicationContext());
        notificationManager.notify(GROUP_KEY, 0, notificationBuilder.build());
    }

    @Override
    public Notification onGenerateNotification(PushMessage pushData) {
        // Summary will be displayed only on devices that do not support grouped notifications
        generateSummary(pushData);

        // This notification will be displayed only on devices that support grouped notifications
        Notification notification = new NotificationCompat.Builder(getApplicationContext())
                .setContentTitle(getContentFromHtml(pushData.getHeader()))
                .setContentText(getContentFromHtml(pushData.getMessage()))
                .setSmallIcon(pushData.getSmallIcon())
                .setTicker(getContentFromHtml(pushData.getTicker()))
                .setWhen(System.currentTimeMillis())
                .setGroup(GROUP_KEY)
                .setAutoCancel(true)
                .build();

        addSound(notification, pushData.getSound());

        return notification;
    }
}

File: Samples/Android-Advanced-Kotlin/app/src/main/java/com/pushwoosh/sample/inbox/InboxNotificationFactory.kt
package com.pushwoosh.sample.inbox

import android.app.Notification
import android.app.PendingIntent
import android.support.v4.app.NotificationCompat
import android.support.v4.app.NotificationManagerCompat
import com.pushwoosh.notification.PushMessage
import com.pushwoosh.notification.PushwooshNotificationFactory

class InboxNotificationFactory : PushwooshNotificationFactory() {
    companion object {
        private const val GROUP_KEY = "PushwooshGroup"
    }

    private fun generateInboxStyle(pushData: PushMessage): NotificationCompat.Style? {
        if (applicationContext == null) {
            return null
        }

        MessageStorage.addMessage(applicationContext!!, Message(pushData.message, System.currentTimeMillis(), pushData.header))
        val messages = MessageStorage.getHistory(applicationContext!!)

        messages?. run {
            val style = NotificationCompat.InboxStyle().setBigContentTitle(pushData.header + " Details")

            for (message in messages) {
                style.addLine(getContentFromHtml("<b>" + message.sender + "</b> " + message.text))
            }

            if (messages.size > 7) {
                style.setSummaryText("+ " + (messages.size - 7) + " more")
            }

            return style
        }

        return null
    }

    private fun generateSummary(pushData: PushMessage) {
        val notificationBuilder = NotificationCompat.Builder(applicationContext)
                .setContentTitle(getContentFromHtml(pushData.header))
                .setContentText(getContentFromHtml(pushData.message))
                .setSmallIcon(pushData.smallIcon)
                .setTicker(getContentFromHtml(pushData.ticker))
                .setWhen(System.currentTimeMillis())
                .setStyle(generateInboxStyle(pushData))
                .setGroup(GROUP_KEY)
                .setGroupSummary(true)
                .setAutoCancel(true)
                .setContentIntent(PendingIntent.getActivity(applicationContext, 0, getNotificationIntent(pushData), PendingIntent.FLAG_CANCEL_CURRENT))

        val notificationManager = NotificationManagerCompat.from(applicationContext!!)
        notificationManager.notify(GROUP_KEY, 0, notificationBuilder.build())
    }

    override fun onGenerateNotification(pushData: PushMessage): Notification? {
        // Summary will be displayed only on devices that do not support grouped notifications
        generateSummary(pushData)

        // This notification will be displayed only on devices that support grouped notifications
        val notification = NotificationCompat.Builder(applicationContext)
                .setContentTitle(getContentFromHtml(pushData.header))
                .setContentText(getContentFromHtml(pushData.message))
                .setSmallIcon(pushData.smallIcon)
                .setTicker(getContentFromHtml(pushData.ticker))
                .setWhen(System.currentTimeMillis())
                .setGroup(GROUP_KEY)
                .setAutoCancel(true)
                .build()

        addSound(notification, pushData.sound)

        return notification
    }

}
